usaspending_api/broker/management/commands/load_fpds_from_broker.py

```diff
@@ -0,0 +1,139 @@
+from django.core.management.base import BaseCommand
+import logging
+import re
+import psycopg2
+from datetime import datetime, timezone, timedelta
+
+from usaspending_api.data_load.fpds_loader import run_fpds_load, destroy_orphans, load_chunk
+from usaspending_api.common.retrieve_file_from_uri import RetrieveFileFromUri
+from usaspending_api.common.helpers.date_helper import datetime_command_line_argument_type
+from usaspending_api.common.helpers.sql_helpers import get_broker_dsn_string
+from usaspending_api.common.helpers.etl_helpers import update_c_to_d_linkages
+from usaspending_api.etl.award_helpers import update_awards, update_contract_awards, update_award_categories
+from usaspending_api.broker.helpers.last_load_date import get_last_load_date, update_last_load_date
+
+logger = logging.getLogger("console")
+
+BROKER_CONNECTION_STRING = get_broker_dsn_string()
+
+CHUNK_SIZE = 5000
+
+ALL_FPDS_QUERY = "SELECT * FROM detached_award_procurement"
```

Comment:
Can you explain how the schema change could break this? It seems like select * would be as stable as any column list

Reply:
It is the old and true explicit vs implicit argument. If Broker's schema broke the loader, wouldn't you wish the SQL fetching the rows to break instead of some potentially vague place down in the loader code?

https://stackoverflow.com/a/3640051

Reply:
I would like it to break in the mappings, at the precise line where the missing column is asked for by either a function or conversion, as a key error in the load object. There are no promises that there isn't some broker column used in one of the derivation functions and not in  field_mappings_fpds.py directly, so any derived list drawn from the code would be tricky to maintain as column logic evolves.

It's possible there's still a benefit to speed if there are columns we aren't using, but that seems like a smaller concern to me.

Reply:
> There are no promises that there isn't some broker column used in one of the derivation functions and not in field_mappings_fpds.py directly.

I wasn't aware that 100% of the fields retrieved from Broker were not in `field_mappings_fpds.py`. Now I think it is a very good idea to specifically list the broker columns in this SQL (as a "document" is you will) so it is easy to see the specific fields necessary for the script.

Reply:
Congrats for making me need a good 15 minutes to come up with an intelligent reply to this. I suppose this gets to the heart of the documentation vs. self-documenting debate.
Here's how I see it: 
- There's a set of columns from broker needed to populate the usaspending tables
- This set can be derived, but only by tracing through all of the functions, and could change as the functions change
If these were listed explicitly:
- It would be possible to see at a glance what columns broker would need to not change in order to avoid breaking the loader, but if a function started using a new column in the background it would still be able to break the loader 
- Broker could still change things unilaterally (and they have no reason to check this code before altering their schema), and it would still break the loader. In fact, now a dev would have to change USASpending's code in two places to fix it instead of just one

Reply:
> ....but if a function started using a new column in the background it would still be able to break the loader
This is exactly how the current loader works, things hidden deeply in code which requires a developer to dig through multiple files to learn what is occurring. No external document is going to "keep the code honest" so to speak. My understanding is a major goal for this refactor was stated along the lines of "things which are wrong will break fast and be easy to fix" (based on memory, please correct if inaccurate). Explicitly setting the columns appears to be in line with everyone's expectations for improving the script.

> Broker could still change things unilaterally (and they have no reason to check this code before altering their schema), and it would still break the loader.
Yes, the changes would break the script and execution would stop immediately. If the DB error is logged it will clearly point to the broken SQL and expected column which is missing. It would be the optimal time to throw an exception.

> In fact, now a dev would have to change USASpending's code in two places to fix it instead of just one
That was why I originally wrote "broker columns are already defined in field_mappings_fpds.py" since I assumed all of the necessary columns lived in dictionaries in that file. That way the SQL is built using the list of fields necessary for the load operations and aren't two separate lists. A developer *should* need to add the field to some mapping for it to be stored in USAspending. If there are necessary fields being retrieved from Broker and not listed in `field_mappings_fpds.py` that needs to change.

Reply:
We are all in agreement that the loader should fail quickly and loudly if the schema breaks. As it stands, we have that in the key error, either in fpds_loader for the standard and boolean columns, or in the function itself for function-based columns. The nice thing about that is that you get to find out exactly what key it was looking for, and which line of code was actually using it for something. 
By adding a guard clause up here, you only give the developer the former, and only if the name is written accurately. If someone made the mistake of misspelling the column name in the column list, tracing what this column was used for becomes guesswork. The same issue arises by duplicating the list in field_mappings_fpds.py. 
Writing out the business logic in plain language is likely valuable documentation, but adding a list of columns with no explanation doesn't accomplish that.

Reply:
I think both of the solutions work here. @tony-sappe suggests an approach that is a bit more explicit, and probably _a **bit** more_ favorable than the way @BrianZito is suggesting to surface the error because it gives an error that points out the root of the problem as a schema-to-query incompatibility. (As noted further down, a similar issue exists on the load/insert side)

If we couldn not just utilize the mapping dictionary to get the columns, I might say writing out each by hand in the SQL query is overly verbose, but the keys of that dict do make a good source for this.

I think one way to look at it is pre-load validation. Many ETL tools will do this on your behalf as part of their execution (thinking about the traditional GUI ones, where you drag left-to-right arrows mapping source to destination flows). They will inspect the source and destination of all the data that will flow through, and throw an early validation error before any data is moved (so no data is harmed). 

I kindof like that idea, where a `validate()` step happens preemptively before load starts, and it is able to use the mapping dictionary to check:
- any table.column I intend to _extract_ ... are in-fact at the source
- any table.column I intend to _load_ ... are in fact at the target

validate source cols:
```
cursor.execute("select {source_cols} from {source_tbl} limit 1".format(source_cols=",".join(mapper_dict.keys()), source_tbl="detached_award_procurement")
```
validate target cols:
```
# Note: assuming here some kind of namedtuple with `col` as an element on the RHS of dicts
cursor.execute("select {target_cols} from {target_tbl} limit 1".format(target_cols=",".join([target.col for target in mapper_dict.values()]), target_tbl="transaction_fpds")
```

Giving a fail-early scenario.

Reply:
@BrianZito where do we stand with this thread?
Seems like there are two decent options:

1. Explicitly request specific columns instead of select * ...
2. Add a quick "pre-load validation" check at the beginning to confirm the necessary columns exist before any loading starts.

Reply:
Looks like this slipped through the cracks somewhere in the file naming. I would rather do option one than two, but that's coming from my bias of thinking that either do more harm than good.